### PR TITLE
[tflite] HardSwish and Tanh quant support by dequantizing to float in Lower.

### PR DESCRIPTION
**Pull request by vuzelac-cadence**
Support for quantized operators by lowering them using simple dequantization and float kernels.
Added test models for Tanh and HardSwish.

test models use float input and reference tensors. These are converted to the model input type on load and compare.

### CODE REVIEW

**Comment by vuzelac-cadence** [comment]
@zrphercule @jfix71 
there seem to be missing support for many tflite quantized models and their testing.
Here, I am adding simple lowering of quantized operators by dequantizing them and executing in float.

**Comment by vuzelac-cadence** [comment]
+ @mciprian13

File: utils/scripts/gen_tflite_models.py [hunk 1]
```diff
@@ -288,21 +304,25 @@ def gen_unary_operator_test(name, type, input_shape):
...
-    save_model(model, name)
+    save_model(model, name, quant)
     # Save data.
     save_tensor(inp_tensor, name + ".inp0")
     save_tensor(out_tensor, name + ".out0")
     # Clear session.
     keras_backend.clear_session()
 
-
+gen_unary_operator_test(name="hardSwish", type="hardSwish", input_shape=(1, 10))
```
Comment (mciprian13) [inline utils/scripts/gen_tflite_models.py:323]: I would prefer renaming this model to "hard_swish" (as you see I wanted to be very consistent in terms of naming for the TFLite models ... I would prefer to have this cleaner than ONNX which is a real mess).

File: utils/scripts/gen_tflite_models.py [hunk 2]
```diff
@@ -47,6 +47,12 @@
 OUT_DIR = os.path.join(os.path.dirname(__file__), "tflite_models")
 
 
+# Generate quantization data based on the input tensor
+representative_images = np.random.randn(1,10).astype(np.float32)
```
Comment (mciprian13) [inline utils/scripts/gen_tflite_models.py:51]: Does the quantization mechanism work for models with multiple inputs/ouputs (e.g. add/sub)?

File: tests/unittests/TFLiteImporterTest.cpp [hunk 3]
```diff
@@ -96,12 +97,27 @@ static void loadAndRunModel(std::string modelName, float maxError = 1e-6) {
   size_t inpIdx = 0;
   for (const auto &inpPH : inputPH) {
     std::string inpFilename = dataBasename + ".inp" + std::to_string(inpIdx++);
-    Tensor *inpT = bindings.get(inpPH);
-    loadTensor(inpT, inpFilename);
+
+    auto phTy = inpPH->getNthResult(0).getType();
+    // Load reference tensor.
```
Comment (mciprian13) [inline tests/unittests/TFLiteImporterTest.cpp:102]: Remove this comment.

File: tests/unittests/TFLiteImporterTest.cpp [hunk 4]
```diff
@@ -96,12 +97,27 @@ static void loadAndRunModel(std::string modelName, float maxError = 1e-6) {
...
+    Type newTy = *phTy;
```
Comment (mciprian13) [inline tests/unittests/TFLiteImporterTest.cpp:103]: This code is hard to read. When the input is float we should load the data as before and when the data needs to be quantized we should read the data into another tensor (say inpFloatT) and quantize it into the final tensor inpT. Quantizing a tensor looks very strange. Same goes for the output path as well.

File: utils/scripts/gen_tflite_models.py [hunk 5]
```diff
@@ -116,6 +122,14 @@ def save_model(model, filename):
         output_arrays=model_outputs_array,
     )
     converter.dump_graphviz_video = False
+    
+    if (quant):
+      converter.optimizations = [tf.lite.Optimize.DEFAULT]
+      converter.representative_dataset = representative_dataset
```
Comment (mciprian13) [inline utils/scripts/gen_tflite_models.py:128]: We should provide this through the `save_model` function API and not through the global variable. Also for multiple model inputs should this be an array?

**Review by mciprian13 — COMMENTED** [review]
Couple of items/questions.

File: tests/unittests/TFLiteImporterTest.cpp (hunk 4 above)
Comment (vuzelac-cadence) [reply tests/unittests/TFLiteImporterTest.cpp:103]: There is no quantization in place. Output tensor is assigned, not pointed to. Please look at quantizeTensor function - temporary tensor is introduced, quantized, and then return.

Comment (vuzelac-cadence) [reply tests/unittests/TFLiteImporterTest.cpp:105]: Added few more comments to explain the proces.

File: utils/scripts/gen_tflite_models.py (hunk 2 above)
Comment (vuzelac-cadence) [reply utils/scripts/gen_tflite_models.py]: This was not my intention to do as I don't needed - at the moment, the quantization mechanism doesn't break anything - it just targets unary ops. However, I added a bit more to this for completness. Added add_int8 model gen. and tested that it works.

File: utils/scripts/gen_tflite_models.py (hunk 5 above)
Comment (vuzelac-cadence) [reply utils/scripts/gen_tflite_models.py]: Agree, there is now a generator that is passed with save_model and also, the generator is made generic - it simply accepts the list of any length. Verifed with unary and binary models only.

**Comment by vuzelac-cadence** [comment]
@hl475

**Comment by hl475** [comment]
We have an on call rotation from Facebook side to review Glow PRs, and the current on call is @chenccfb . I will let him take a look :)

File: lib/Optimizer/Lower/Lower.cpp [hunk 6]
```diff
@@ -1906,7 +1955,8 @@ bool glow::lowerNode(Function *F, Node *node, CompilationContext &cctx) {
     CASE_LOWER(LSTMUnit);
     CASE_LOWER(Broadcast);
     CASE_LOWER(LogSoftMax);
-    CASE_LOWER(HardSwish);
+    CASE_DEQUANT(Tanh, ElemKind::FloatTy);
+    CASE_LOWER_QUANT(HardSwish, ElemKind::FloatTy);
```
Comment (jfix71) [inline lib/Optimizer/Lower/Lower.cpp:1959]: One concern I have here is if we have fp16 conversion turned on. Lowering happens after fp16 conversion. So we'd end up with some ops such as Tanh/HardSwish that will be in fp32 even though fp16 conversion would have liked to convert them to fp16. Any thoughts here?

Comment (jfix71) [inline lib/Optimizer/Lower/Lower.cpp:1959]: We have hard swish lowering implemented at this point in master. Can you rebase?

**Review by jfix71 — COMMENTED** [review]
I'm not convinced that we want to handle this `dequant -> node -> quant` conversion through lowering. It makes it a bit weird to handle in terms of a backend requesting something not be lowered. It's more like what we have in https://github.com/pytorch/glow/pull/5122 (which has kind of stalled out, not sure if/when it'll land) -- I wonder if we should do something similar instead, or see if we can get that landed, which I think would also solve this problem? WDYT?

**Comment by vuzelac-cadence** [comment]
@jfix71, Lowering happens before these optimizations, by the time something can be handled in optimization passes the node is already not there, if lowered, thus there is no node anymore to handle for optimization.

Or, from the lowering perspective: pickup BatchNorm for example. That lowering is not going to work - it needs float implementation. That was the reason why I added dequant here - like an extension to make existing lowering to work. In that light, only nodes that rely on float point lowering type are to be here.

**Comment by jfix71** [comment]
@vuzelac-cadence Sorry for the delay. I am still not super happy about doing this as a lowering pass. It makes sense to an extent with the philosophy of the lowering pass, however in general we tend to rely on lowering for when an op is supported at all based on its kind, and not based on its precision. This PR blurs the lines a bit more, for example it will require a backend to return false for quantized HardSwish and Tanh in `Backend::shouldLower()`, even though it returns true for them in `Backend::isOpSupported()`. I would rather not add more complex interactions between these two APIs if we can help it.

As mentioned, https://github.com/pytorch/glow/pull/5122 is a more generalized implementation for this support. It's currently stalled out/author hasn't responded in a couple months. So if you want this to land beforehand, perhaps you could add a specialized call inside the graph optimization pipeline similar to #5122 that does this lowering based on whether `Backend::isOpSupported()` returns true for quantized HardSwish/Tanh. Then when https://github.com/pytorch/glow/pull/5122 lands we could remove your specialized pass to allow the general one to do the same thing. Or you could just commandeer https://github.com/pytorch/glow/pull/5122 and land it yourself 😛
